lib/std/mem.zig

```diff
@@ -630,6 +631,140 @@ test "spanZ" {
+/// and no sentinel is found, the full length of the array/slice is returned.
+pub fn scanZ(ptr: anytype) Span(@TypeOf(ptr)) {
+    if (@typeInfo(@TypeOf(ptr)) == .Optional) {
+        if (ptr) |non_null| {
+            return spanZ(non_null);
+        } else {
+            return null;
+        }
+    }
+    const Result = Span(@TypeOf(ptr));
+    const l = scanLenZ(ptr);
+    if (@typeInfo(Result).Pointer.sentinel) |s| {
+        return ptr[0..l :s];
+    } else {
+        return ptr[0..l];
+    }
+}
+
+test "scanZ" {
+    const testing = std.testing;
+    testing.expectEqualSlices(u8, "aoeu", scanZ("aoeu"));
+
+    {
+        var array: [5]u16 = [_]u16{ 1, 2, 3, 4, 5 };
+        testing.expectEqualSlices(u16, &array, scanZ(&array));
+        testing.expectEqualSlices(u16, array[0..3], scanZ(array[0..3]));
+        array[2] = 0;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(&array));
+        testing.expectEqualSlices(u16, array[0..2], scanZ(array[0..3]));
+        const sentinel_ptr = @as([*:0]u16, array[0..2 :0]);
+        testing.expectEqualSlices(u16, array[0..2], scanZ(sentinel_ptr));
+        const c_ptr = @as([*c]u16, &array);
+        testing.expectEqualSlices(u16, array[0..2], scanZ(c_ptr));
+        const slice: []u16 = &array;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(slice));
+        const sentinel_slice: [:0]u16 = array[0..2 :0];
+        testing.expectEqualSlices(u16, array[0..2], scanZ(sentinel_slice));
+    }
+    {
+        var array: [5:0]u16 = [_:0]u16{ 1, 2, 3, 4, 5 };
+        testing.expectEqualSlices(u16, &array, scanZ(&array));
+        array[2] = 0;
+        testing.expectEqualSlices(u16, array[0..2], scanZ(&array));
+    }
+}
+
+/// Private helper for scanZ. If you want the scanned length, use scanZ(foo).len
+fn scanLenZ(ptr: anytype) usize {
+    switch (@typeInfo(@TypeOf(ptr))) {
+        .Array => |info| {
+            if (info.sentinel) |sentinel| {
+                if (sentinel != 0) {
+                    @compileError("std.mem.scanZ requires the sentinel, if any, to be 0");
+                }
+                return indexOfSentinel(info.child, sentinel, &ptr);
+            } else {
+                return indexOfScalar(info.child, &ptr, 0) orelse info.len;
+            }
+        },
+        .Pointer => |info| switch (info.size) {
+            .One => switch (@typeInfo(info.child)) {
+                .Array => |x| {
+                    if (x.sentinel) |sentinel| {
+                        if (sentinel != 0) {
+                            @compileError("std.mem.scanZ requires the sentinel, if any, to be 0");
+                        }
+                        return indexOfSentinel(x.child, sentinel, ptr);
+                    } else {
+                        return indexOfScalar(x.child, ptr, 0) orelse x.len;
+                    }
+                },
+                else => @compileError("invalid type given to std.mem.scanZ"),
```

Comment:
Since each handled case does `return`, you could simplify this by using `else => {}` here and having this compile error only once after the outer `switch` body.